src/iframe-helper.js

```diff
@@ -17,6 +17,77 @@
 import {dev} from './log';
 import {parseUrl} from './url';
 
+/**
+ * @typedef {{
+ *   win: !Window,
+ *   events: !Object<string, !Array<function(!Object)>>
+ * }}
+ */
+var WindowEventsDef;
+
+/**
+ * Returns a mapping from a URL's origin to an array of windows and their listenFor listeners.
+ * @param {!Window} parentWin the window that created the iframe
+ * @param {boolean} opt_create create the mapping if it does not exist
+ * @return {?Object<string, !Array<!WindowEventsDef>>}
+ */
+function getListenFors(parentWin, opt_create) {
+  let listeningFors = parentWin.listeningFors;
+
+  if (!listeningFors && opt_create) {
+    listeningFors = parentWin.listeningFors = Object.create(null);
+  }
+  return listeningFors || null;
+}
+
+/**
+ * Returns an array of WindowEventsDef that have had any listenFor listeners registered for this origin.
+ * @param {!Window} parentWin the window that created the iframe
+ * @param {string} origin the child window's origin
+ * @param {boolean} opt_create create the array if it does not exist
+ * @return {?!Array<!WindowEventsDef>}
+ */
+function getListenForOrigin(parentWin, origin, opt_create) {
+  const listeningFors = getListenFors(parentWin, opt_create);
+  if (!listeningFors) {
+    return listeningFors;
+  }
+
+  let listenOrigin = listeningFors[origin];
+  if (!listenOrigin && opt_create) {
+    listenOrigin = listeningFors[origin] = [];
+  }
+  return listenOrigin || null;
+}
+
+/**
+ * Returns an mapping of event names to listenFor listeners.
+ * @param {!Window} parentWin the window that created the iframe
+ * @param {string} origin the child window's origin
+ * @param {!Window} listenForWin the child window
+ * @param {boolean} opt_create create the mapping if it does not exist
+ * @return {?Object<string, !Array<function(!object)>>}
+ */
+function getListenForEvents(parentWin, origin, listenForWin, opt_create) {
+  const listenOrigin = getListenForOrigin(parentWin, origin, opt_create);
+
+  if (!listenOrigin) {
+    return listenOrigin;
+  }
+  let windowEvents = listenOrigin.find(we => {
+    return we.win === listenForWin;
+  });
+
+  if (!windowEvents && opt_create) {
+    windowEvents = {
+      win: listenForWin,
```

Comment:
I'm not sure how much of a memory leak this can turn into. Should I store `iframe` instead?